### ISSUE TITLE
HB not using correct CLT from the booted macOS if machine has multiple macOS systems with old Xcode versions

### ISSUE THREAD

**Issue by MarioG-X** [Homebrew/brew#10745]
**`brew config` output**

THIS IS WITH OTHER MACOS PARTITIONS MOUNTED (BAD - Using Xcode from different system instead of local CLT):
HOMEBREW_VERSION: 3.0.2
ORIGIN: https://github.com/Homebrew/brew
HEAD: b712bcc01e8fdaae25d6f232f51637bf00db307f
Last commit: 6 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 14b0719d3f7119bae97ea7dd9f0f4dd3ac9cd159
Core tap last commit: 25 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: 12.1 => /Volumes/FX 10.15.7 Xcode/Applications/Xcode.app/Contents/Developer. <- Note this was the only volume with full Xcode, others had only CLT, but other wrong volumes would be used if this one was unmounted.

THIS IS WITH ALL OTHER MACOS PARTITIONS DISMOUNTED (GOOD - uses CLT from booted system):
OMEBREW_VERSION: 3.0.2
ORIGIN: https://github.com/Homebrew/brew
HEAD: b712bcc01e8fdaae25d6f232f51637bf00db307f
Last commit: 6 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: d2b380baa4e53195e89e90a146d87ef08c529b1a
Core tap last commit: 20 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A

**`brew doctor` output**

THIS IS WITH OTHER MACOS PARTITIONS MOUNTED (BAD - Using Xcode from different system instead of local CLT):
Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

Warning: Xcode is installed to a directory with a space in the name.    <- NOT TRUE
This will cause some formulae to fail to build.

Warning: Your Xcode (12.1) is outdated.
Please update to Xcode 12.4 (or delete it).
Xcode can be updated from the App Store.

THIS IS WITH ALL OTHER MACOS PARTITIONS DISMOUNTED (GOOD - uses CLT from booted system):
Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

****

- [X] The `brew doctor` above contains no "Warning" lines.

**What were you trying to do (and why)?**

Brew upgrade with some needing config and make. The machine has 9 macOSs on it because it is used to weekly test updates for Freeswitch. I maintain the macOS section of the wiki and test the last 3 supported macOSs. For now Mohave, Catalina and Big Sur. There may be multiple versions fo each for debugging. They will have different levels of CLT, for instance Mojave has version 11, while the other two, different levels of version 12. This has caused many days of problem diagnosis. If you don't look for the message, there may be output differences, if you're lucky, make will fail.

If you look at the brew config and doctor, they are different depending on if all other volumes are dismounted. Also, since CLT is on 8 volumes, there is no telling which one it will use. I think HB should use the CLT on the booted system and ignore others unless override somehow.

**What happened (include all command output)?**

Warning: Your Xcode (12.1) is outdated.
Please update to Xcode 12.4 (or delete it).   <- this booted system has 12.4, clang -v shows see below
Xcode can be updated from the App Store.

Running pkgutil --pkg-info=com.apple.pkg.CLTools_Executables shows:
package-id: com.apple.pkg.CLTools_Executables
version: 12.4.0.0.1.1610135815
volume: /
location: /
install-time: 1614445498
groups: com.apple.FindSystemFiles.pkg-group

clang -v shows:
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

**What did you expect to happen?**

This message should not occur since the running system had CLT 12.4 installed. I do not use the full Xcode, only CLT downloaded via Xcode-select --install.

**Step-by-step reproduction instructions (by running `brew` commands)**

brew upgrade. <- this is all, if a package requires make and other partitions are mounted the make sometimes fails due to using a random CLT version.

**Comment by Bo98** [comment]
Xcode.app detection is currently done like this:
* Use `/Applications/Xcode.app`, if it exists
* Use what is returned by `mdfind "kMDItemCFBundleIdentifier == com.apple.dt.Xcode"`.

**Comment by MarioG-X** [comment]
Since Xcode is on only 1 of the 9 volumes it picked that up even though it could be a completed wrong version.

So the fix for HB should be to add a search for CLT before using mdfind?
1. Use /Applications/Xcode.app, if it exists
2. Use /Library/Developer/CommandLineTools/. . . .something, if it exists
3. Use what is returned by mdfind "kMDItemCFBundleIdentifier == com.apple.dt.Xcode"
4. Use what is returned by mdfind "kMDItemCFBundleIdentifier == com.apple.pkg.CLTools_Executables"

**Comment by Bo98** [comment]
CLT and Xcode detection is completely separate. Note how the CLT is correctly detected in both `brew config` examples you provided.

> CLT: 12.4.0.0.1.1610135815

In most cases, CLT is the main thing that matters. Xcode is however required for a select few formula, such as `qt`.

**Comment by MarioG-X** [comment]
I think I am not fully clear, there is a problem it during make it uses the wrong partition tools. The output of the two configure are different. The bad one has **Xcode: 12.1**, the good one **Xcode: N/A**.
Even though both have CLT: 12.4.0.0.1.1610135815, the **bad run is not using the correct CLT**, it's using Xcode from a different partition.
This should be fixed, it may cost someone days of unnecessary work like it did for me.

**Comment by Bo98** [comment]
> **bad run is not using the correct CLT**

Given `brew` only ever checks `/Library/Developer/CommandLineTools` for the CLT, I don't know what you mean by this.

**Comment by MarioG-X** [comment]
If brew only ever checks /Library/Developer/CommandLineTools then it's not working.

**Comment by Bo98** [comment]
You can also see `brew --env` which shows exactly what is being used for formula builds.

**Comment by MarioG-X** [comment]
Doesn't solve the problem. It should be fixed.

**Comment by Bo98** [comment]
> CLT on the booted volume HB should look for an older Xcode. 

`brew` always looks for _both_ CLT and Xcode. I can see the argument of fixing the Xcode searching so be limited to one volume which will fix the Xcode part of your issue - I just don't yet see anything to show that the CLT part specifically is broken as claimed.

**Comment by carlocab** [comment]
Unfortunately, it's not all clear what exactly the problem is, and what needs to be done to fix it. It also seems extremely likely that a fix targets such a niche use-case that it's unlikely to be worth prioritising over many other things that need fixing in Homebrew.

What formulae are you building that the misspecified Xcode is causing you problems?

Note that for the vast majority of formulae in homebrew/core, it doesn't matter whether you have an Xcode installation or not. Thus, if you have the latest CLT installed, it almost never matters what Xcode you have lying around, be it Xcode 1.0 or 12.4.

**Comment by MarioG-X** [comment]
Building Freeswitch requires prerequisites that installs dozens of HB packages, could be 80-90 and they change. I do this weekly for three systems on the same machine. The problem is as I stated, the build/make will either cause execution problems (one example: https://github.com/signalwire/freeswitch/issues/961) or if I am lucky fails. This has caused problems over the last year that cost me weeks to track down. I don't see how no one can understand this problem. If you have Xcode or CLT installed on a system, and you run HB, those should be used first before searching for some random Xcode on another volume.

Here is the command I use which have a ton of prerequisites:
brew install autoconf automake curl ffmpeg jpeg ldns libpq libsndfile libtiff libtool lua openssl opus pcre pkg-config speex speexdsp sqlite yasm signalwire/homebrew-signalwire/flite signalwire/homebrew-signalwire/libks signalwire/homebrew-signalwire/signalwire-c signalwire/homebrew-signalwire/sofia-sip signalwire/homebrew-signalwire/spandsp

**Comment by carlocab** [comment]
None of these formulae even have a dependency on Xcode, so it shouldn't matter what versions of Xcode `brew` finds as long as you have a CLT installation.

I guess, to put it differently so that you can understand why I don't understand: you seem convinced that the solution to your problems is for `brew` to look for Xcode installations in some sort of "correct" order. I doubt this actually fixes any of your problems because `brew` shouldn't be using any Xcode installations at all. The order in which they are searched does not matter. We can switch up this order all we want, and it's not going to change a thing for you.

On the off chance that it does change something, then that means something else went horribly wrong. That something else should be fixed instead rather than papered over with a carousel of Xcode search paths. Indeed, even the issues you've linked to show no evidence of having been caused by an Xcode installation -- instead, they seem to be due to issues with the CLT.

However, if the fix is in fact as simple as you claim it is, then you should give the fix a try locally and see if it helps you. You can open a pull request if it does.

**Comment by MikeMcQuaid** [comment]
> However, if the fix is in fact as simple as you claim it is, then you should give the fix a try locally and see if it helps you. You can open a pull request if it does.

Agreed. The current behaviour is by design and we'll review pull requests to improve it.

**Comment by MarioG-X** [comment]
I don't know how to code this stuff let alone HB internals so my comment at choosing a sequence was based on Bo98 previous comments about the order HB uses. It may be another issue especially since you said "brew shouldn't be using any Xcode installations at all", this sound like the problem to me. Anyway, this presents an error in building for me that I can get around by making sure all other volumes on the machine are unmounted before using HB. Since this was closed looks like I will have to just live with the problem, thank goodness I found it and a workaround. Hopefully, this post will help someone else save some time. Thanks for the comments and quick responses.
